### ISSUE TITLE
【FR】Bug: sort operator hangs in torch.compile mode with top_p sampling

### ISSUE THREAD

**Issue by Lxiparer** [FlagOpen/FlagGems#5998]
### Description

When running model inference on model **mlabonne/NeuralDaredevil-8B-abliterated**, the `sort` operator hangs indefinitely during `torch.compile` optimization phase when using `top_p` sampling strategy.

The affected operators include:
- `sort`
- `sort_stable`
- `cumsum` (potentially)

### Environment

| Item | Value |
|------|-------|
| Hardware | 天数 Iluvatar |
| GPU | BI-V150 × 16 |
| OS | Linux |
| Python | 3.10+ |
| PyTorch | 2.1.0+ (with torch.compile support) |
| FlagGems | 5.0.0 |
| FlagTree | 0.6.0 |
| vLLM | 0.20.2 |
| vllm-plugin-FL | 0.2.0 |

### Steps to Reproduce

1. Set up environment on 天数 Iluvatar BI-V150
2. Install dependencies (PyTorch + Triton + FlagGems 5.0.0 + vLLM 0.20.2)
3. Start vLLM inference service with FlagGems enabled and torch.compile
   ```bash
   USE_FLAGGEMS=1 vllm serve mlabonne/NeuralDaredevil-8B-abliterated \
     --tp 8 --port 8000 --enforce-eager=False  # enable torch.compile
   ```
4. Configure sampling with `top_p=0.9`
5. Observe service hangs during torch.compile stage, specifically when processing `sort` operator

### Expected Behavior

The `sort` operator should complete compilation within reasonable time (< 60 seconds) and the service should start successfully with torch.compile optimization enabled.

### Actual Behavior

**Symptom**: Service startup hangs indefinitely at torch.compile stage

**Hang point**: When compiling the top_p sampling code path that uses `sort` operator

**Both V1 (native) and V2 (FlagGems) affected**: Confirms this is a platform-level issue with Iluvatar + torch.compile interaction

**Workaround that works**:
```bash
# Option 1: Disable torch.compile
vllm serve model --enforce-eager

# Option 2: Disable sort operators
export FLAGGEMS_DISABLE_OPS="sort,sort_stable,cumsum"
vllm serve model
```

After applying workaround, service starts successfully but with:
- 20-30% performance loss (compile optimization disabled)
- Limited sampling strategies (top_p functionality degraded)

### Error Logs

```
[INFO] Starting vLLM with torch.compile enabled
[INFO] Compiling model graph...
[DEBUG] Processing top_p sampling code
[DEBUG] Compiling sort operation for logits sorting...
(hangs indefinitely - no further output)

# After timeout/manual kill:
[ERROR] torch.compile阶段停滞
[ERROR] FlagGems和native均在torch.compile阶段停滞
[ERROR] 无法归因具体算子
[WARNING] Service startup aborted after timeout
```

### FlagGems Integration Code

**top_p sampling workflow**:
```python
# In vLLM sampler.py
def top_p_sampling(logits, p=0.9):
    # 1. Convert to probabilities
    probs = torch.softmax(logits, dim=-1)
    
    # 2. Sort probabilities (HANG POINT)
    sorted_probs, sorted_indices = torch.sort(probs, descending=True)
    
    # 3. Cumulative sum
    cumsum_probs = torch.cumsum(sorted_probs, dim=-1)
    
    # 4. Find cutoff
    mask = cumsum_probs <= p
    
    # 5. Sample from filtered distribution
    ...
```

**torch.compile context**:
```python
@torch.compile
def sampling_step(logits):
    return top_p_sampling(logits, p=0.9)
```

**FlagGems enable call**:
```python
import flag_gems
flag_gems.enable()
```

### Additional Context

- **Affects**: Models using top_p/nucleus sampling with torch.compile
- **Platform**: Iluvatar BI-V150 specific (NVIDIA platforms work normally)
- **Operator behavior**:
  - `eager` mode: ✅ sort works correctly
  - `compile` mode: ❌ sort hangs indefinitely
- **Confirmed on**: 0810FlagRelease batch migration testing
- **Recovery**: success (through disabling sort operators)
- **Known issue**: Reported as `top_p=0.9 采样导致 sort 算子挂起`

**Tested workarounds**:
- ✅ `--enforce-eager`: Works but loses compile optimization
- ✅ `FLAGGEMS_DISABLE_OPS=sort`: Works but limits sampling strategies
- ❌ Alternative sort implementations: Not available

### Possible Directions

1. **Short-term: Add timeout + automatic fallback** (Recommended for quick fix)
   ```python
   import signal
   
   def compile_with_timeout(fn, timeout=60):
       def timeout_handler(signum, frame):
           raise TimeoutError("torch.compile timeout")
       
       signal.signal(signal.SIGALRM, timeout_handler)
       signal.alarm(timeout)
       try:
           return torch.compile(fn)
       except TimeoutError:
           logger.warning("torch.compile timeout, fallback to eager")
           return fn  # fallback to eager mode
       finally:
           signal.alarm(0)
   ```

2. **Mid-term: Implement compile-safe sort for Iluvatar**
   - Investigate why sort hangs in compile mode on Iluvatar
   - Add platform detection: `if is_iluvatar() and torch.compiler.is_compiling()`
   - Use simpler sort implementation or disable fusion

3. **Alternative: top_p without sort**
   - Use `torch.topk` approximation instead of full sort
   - Trade slight quality loss for compilation stability
   ```python
   def top_p_via_topk(logits, p=0.9):
       k = estimate_k(p, vocab_size)  # heuristic
       topk_probs, topk_indices = torch.topk(probs, k)
       # sample from topk
   ```

4. **Long-term: Root cause analysis**
   - Debug Iluvatar Triton backend interaction with torch.compile
   - Submit bug report to Iluvatar SDK team
   - Collaborate with PyTorch torch.compile team

**Estimated effort**: 
- Timeout fallback: 0.5 day
- Root cause fix: 1-2 weeks

**Priority**: P1 (has workaround, but impacts performance and functionality)

---

**Reporter**: FlagRelease Automation  
**Date**: 2026-09-04  
**Affected models**: mlabonne/NeuralDaredevil-8B-abliterated  
**Month**: 0810FlagRelease  
**Status**: Temporarily mitigated via `--enforce-eager`
